### PR TITLE
CXX-531 Test agains all sources of stdx polyfills

### PR DESCRIPTION
**Pull request by tessavitabile**
https://evergreen.mongodb.com/version/5697cd8a3ff122756500022b_0

### CODE REVIEW

File: .mci.yml [hunk 1]
```diff
@@ -24,6 +24,8 @@ variables:
     ## cmake flag variables
     cmake_flags:
         ubuntu_cmake_flags: &ubuntu_cmake_flags -DCMAKE_CXX_FLAGS="-Wall -Wextra -Wno-attributes -Werror -Wno-error=missing-field-initializers"
+        ubuntu_boost_cmake_flags: &ubuntu_boost_cmake_flags -DBSONCXX_POLY_USE_BOOST=ON -DCMAKE_CXX_FLAGS="-Wall -Wextra -Wno-attributes -Werror -Wno-error=missing-field-initializers"
```
Comment (acmorrow) [inline .mci.yml]: Can you add `-DBSONCXX_POLY_USE_MNMLSTC=OFF` here as well? Currently, the options aren't well factored (I have something in flight) but for now it is the right thing to do.

File: .mci.yml [hunk 2]
```diff
@@ -24,6 +24,8 @@ variables:
...
+        ubuntu_std_experimental_cmake_flags: &ubuntu_std_experimental_cmake_flags -DBSONCXX_POLY_USE_STD_EXPERIMENTAL=ON -DCMAKE_CXX_STANDARD=14 -DCMAKE_CXX_FLAGS="-Wall -Wextra -Wno-attributes -Werror -Wno-error=missing-field-initializers"
```
Comment (acmorrow) [inline .mci.yml]: Here as well for explicitly disabling MNMLSTC as well as enabling std::experimental

**Comment by acmorrow** [comment]
@tessavitabile FYI if #435 goes first, you don't need to change anything. So, LGTM if we sequence things right.

File: .mci.yml [hunk 3]
```diff
@@ -151,6 +153,32 @@ buildvariants:
       tasks:
           - name: compile_and_test
 
+    - name: ubuntu1404-debug-boost
+      display_name: "Ubuntu 14.04 Debug with boost"
```
Comment (samantharitter) [inline .mci.yml]: for brevity, maybe just "Ubuntu 14.04 Debug (boost)" and "Ubuntu 14.04 Debug (experimental)".  In some views mci doesn't show long names well.

I'd also be ok with an even shorter version, like "Ubuntu (boost)" and "Ubuntu (experimental)."  Of course with this you'd have to go look in .mci.yml to see the ubuntu version, but meh.  I leave it to your discretion.

**Comment by samantharitter** [comment]
Nit about long buildvariant names, otherwise LGTM

**Comment by tessavitabile** [comment]
When I added -DBSONCXX_POLY_USE_MNMLSTC=OFF, I had compile errors. For boost, it could not find boost/smart_ptr/make_unique.hpp. For std experimental, it could not find experimental/string_view.

https://evergreen.mongodb.com/version/5697d4a43ff12275650002de_0

**Comment by acmorrow** [comment]
Interesting. I've tested them all locally. I think without the options refactor it wasn't actually honoring the request to use boost or std::experimental and was still using mnmlstc.

**Comment by tessavitabile** [comment]
Okay. After #435 is committed, I'll remove -DBSONCXX_POLY_USE_MNMLSTC=OFF, rebase and test on evergreen again.

**Comment by acmorrow** [comment]
Well...

- For boost, it looks like the boost is too old to have support for `make_unique`. Can you find out what the first version of boost to offer that was? Then you can bump the minimum required boost in the `find_package(Boost` call in the bsoncxx CMakeLists.txt file.
- For std::experimental, it looks like GCC 4.8 doesn't offer it. Want to see if we get better mileage with clang and libc++? I'm not sure that is installed either. Maybe try it on OS X?

**Comment by acmorrow** [comment]
FYI #435 is in, so you shouldn't need to make any changes per my earlier comments. The lack of building though may still be an issue :)

**Comment by tessavitabile** [comment]
For boost, the first version including make_unique was 1.56.0. The ubuntu 14.04 machines have 1.54.0. However, our OSX builder is already using boost. Is that a sufficient test?

For std::experimental, I tried using clang and libc++ on ubuntu, but cmake considers that clang insufficient. I tried it on OSX and got the same error: that experimental/string_view is not found. I could try clang++ on ubuntu?

**Comment by acmorrow** [comment]
Well, first thing, lets add the 1.58 requirement to the `find_package(Boost...)` so that we lock that knowledge in. That will help prevent users from getting confused.

As far as the OS X builder, I don't think it was actually using boost:

https://evergreen.mongodb.com/task_log_raw/mongodb_cpp_driver_dev_osx_108_compile_and_test_de649bf1acba78db9f744c99f36a3655b540a2eb_16_01_13_23_26_21/0?type=T

And look for EP_. However, my change to the polyfill options means that it will start using boost Real Soon Now (as soon as the build queue catches up). So, we will find out.

We may just not have a new enough system available on MCI to test with std::experimental at this time. Hopefully we will at some point soon when we upgrade the OS X builders, or introduce new variants.

**Comment by tessavitabile** [comment]
Oh, I see. (Though I wonder why turning on Boost caused it to compile on OS X, if it wasn't actually using Boost? Very confusing.)

File: .mci.yml (hunk 1 above)
Comment (acmorrow) [inline .mci.yml]: Please disregard since the other commit merged.

File: .mci.yml (hunk 2 above)
Comment (acmorrow) [inline .mci.yml]: Ditto

**Comment by tessavitabile** [comment]
Should I push the boost minimum in a minor commit and abandon this for now?

File: src/bsoncxx/CMakeLists.txt [hunk 4]
```diff
@@ -90,7 +90,7 @@ include_directories(
 link_directories(${LIBBSON_LIBRARY_DIRS})
 
 if (BSONCXX_POLY_USE_BOOST)
-    find_package(Boost REQUIRED)
+    find_package(Boost 1.56.0 REQUIRED)
```
Comment (acmorrow) [inline src/bsoncxx/CMakeLists.txt]: Thanks

File: .mci.yml (hunk 2 above)
Comment (acmorrow) [inline .mci.yml]: According to this:

http://stackoverflow.com/questions/30995705/cplusplus-201402l-return-true-in-gcc-even-when-i-specified-std-c14

We need either GCC-5.1 or clang-3.5.

I think you should check in with Jonathan and see if we have /any/ MCI images that offer these toolchains yet.

**Comment by acmorrow** [comment]
@tessavitabile - Yes, please push the boost minimum commit as is under this ticket. We may need to hold off on the reset.

File: .mci.yml (hunk 2 above)
Comment (acmorrow) [inline .mci.yml]: Great news: Our Ubuntu 14.10 variant has clang-3.5.

Comment (tessavitabile) [inline .mci.yml]: Indeed! Sorry, I should have commented. It has std::experimental, but something is broken, which I was looking at:
https://evergreen.mongodb.com/task_log_raw/mongodb_cpp_driver_dev_ubuntu1404_debug_std_experimental_compile_and_test_3f3499ee80b26fe3c110340520b959dbf7df846a_16_01_14_21_52_21/0?type=T

Comment (acmorrow) [inline .mci.yml]: And it looks like libc++ on that platform has experimental optional and string_view. I'm trying to build on a spawnhost right now.

**Comment by acmorrow** [comment]
We are going to get testing with boost on Windows once these tickets are done:

https://jira.mongodb.org/browse/BUILD-1129
https://jira.mongodb.org/browse/BUILD-1136

It appears that std::experimental just isn't ready yet given our current builders. Hopefully that will not be an issue once @jbreams new toolchain is up and running, as that will give us GCC 5.3

**Comment by acmorrow** [comment]
@tessavitabile I think we should close this PR if we don't plan to do more with it right now.

**Pull request closed without merging**